### PR TITLE
Stage3

### PR DESCRIPTION
**Pull request by barvered96**

### CODE REVIEW

**Comment by yardenfi** [comment]
Remove e2e directory. You won't need it

File: src/app/app.component.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { Component } from '@angular/core';
+
+@Component({
+  selector: 'app-root',
+  templateUrl: './app.component.html',
+  styleUrls: ['./app.component.css']
+})
+export class AppComponent {
+  title = 'Swagger';
+  constructor() {
```
Comment (yardenfi) [inline src/app/app.component.ts]: remove

File: src/app/create-project/create-project.component.css [hunk 2]
```diff
@@ -0,0 +1,4 @@
+.forming{
```
Comment (yardenfi) [inline src/app/create-project/create-project.component.css]:
```suggestion
.forming {
```

Comment (yardenfi) [inline src/app/create-project/create-project.component.css]: css classes are usually nouns, not actions.

File: src/app/homepage/homepage.component.html [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<div>
```
Comment (yardenfi) [inline src/app/homepage/homepage.component.html]: why is this needed?

File: src/app/header/header.component.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import { MatSidenav } from '@angular/material/sidenav';
+
+@Component({
+  selector: 'app-header',
+  templateUrl: './header.component.html',
+  styleUrls: ['./header.component.css']
+})
+export class HeaderComponent implements OnInit {
+  @Input() sideNav: MatSidenav;
+  constructor() { }
```
Comment (yardenfi) [inline src/app/header/header.component.ts]: remove

File: src/app/header/header.component.html [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+          <a class="nav-link" href="/createProject">Create project <span class="sr-only"></span></a>
+        </li>
+        <li class="nav-item">
+          <a class="nav-link" href="/createResource">Create resource <span class="sr-only"></span></a>
+        </li>
+        <li class="nav-item">
+          <a class="nav-link" href="/viewProjects"> View your projects <span class="sr-only"></span></a>
+        </li>
+      </ul>
+
```
Comment (yardenfi) [inline src/app/header/header.component.html]:
```suggestion
```

File: src/app/header/header.component.css [hunk 6]
```diff
@@ -0,0 +1,8 @@
+section {
+  display: table;
+  width:100%;
```
Comment (yardenfi) [inline src/app/header/header.component.css]:
```suggestion
  width: 100%;
```

File: src/app/dialog/dialog.component.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+      description: this.description,
+      action: this.action,
+      delete_flag: true
+    };
+    this.onClose.next(resource);
+    this.ref.hide();
+  }
+
+  public onCancel(): void {
+    console.log('h');
```
Comment (yardenfi) [inline src/app/dialog/dialog.component.ts]: remove

File: src/app/dialog/dialog.component.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+export class DialogComponent implements OnInit {
+  public onClose: Subject<any>;
+  public name: string;
+  public api: string;
+  public description: string;
+  public resources: Resource[];
+  public action: string;
+  constructor(public ref: BsModalRef<DialogComponent>) {}
+  ngOnInit(): void {
+    this.onClose = new Subject();
```
Comment (yardenfi) [inline src/app/dialog/dialog.component.ts]: Why isn't there a type in this Subject generic? (Subject<T>)

File: src/app/dialog/dialog.component.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    this.ref.hide();
+  }
+
+  public onSaveResource(): void {
+    const resource = {
+      name: this.name,
+      api: this.api,
+      description: this.description,
+      action: this.action,
+      delete_flag: false
```
Comment (yardenfi) [inline src/app/dialog/dialog.component.ts]: js object keys are camelCase

File: src/app/interfaces/base-model.ts [hunk 10]
```diff
@@ -0,0 +1,5 @@
+export interface BaseModel {
+  name: string;
+  api: string;
```
Comment (yardenfi) [inline src/app/interfaces/base-model.ts]: "api" isn't really descriptive... What does this mean?

File: package.json [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  },
+  "private": true,
+  "dependencies": {
+    "@angular/animations": "^11.0.8",
+    "@angular/cdk": "^11.0.3",
+    "@angular/common": "~11.0.6",
+    "@angular/compiler": "~11.0.6",
+    "@angular/core": "~11.0.6",
+    "@angular/forms": "~11.0.6",
+    "@angular/material": "^11.0.3",
```
Comment (yardenfi) [inline package.json]: Do you still need this?

File: src/app/material.module.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
+import { NgModule } from '@angular/core';
+import {MatToolbarModule} from '@angular/material/toolbar';
+import { MatIconModule } from '@angular/material/icon';
+import {MatButtonModule} from '@angular/material/button';
+import {MatDividerModule} from '@angular/material/divider';
+import {MatMenuModule} from '@angular/material/menu';
+import {MatCardModule} from '@angular/material/card';
```
Comment (yardenfi) [inline src/app/material.module.ts]: Do you still need this file?

File: src/app/dialog/dialog.component.html [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+      </div>
+      <div class="modal-body">
+        <form class="row g-3">
+          <div class="col-12">
+            <label class="form-label">Entity Name: </label>
+            <input name="name" type="text" [(ngModel)]="name">
+          </div>
+          <div class="col-12">
+            <label class="form-label">API url: </label>
+            <input name="api" type="text" [(ngModel)]="api">
```
Comment (yardenfi) [inline src/app/dialog/dialog.component.html]: missing bootstrap form classes
https://getbootstrap.com/docs/4.0/components/forms/

File: src/app/dialog/dialog.component.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
+import {Component, OnInit} from '@angular/core';
```
Comment (yardenfi) [inline src/app/dialog/dialog.component.ts]: This component has generic aspects so that you can use it in multiple places, but it also has a ton of edge cases and coupling. Either make it generic with little to none of the coupling, or split it to multiple components

File: src/app/interfaces/base-model.ts [hunk 15]
```diff
@@ -0,0 +1,5 @@
+export interface BaseModel {
```
Comment (yardenfi) [inline src/app/interfaces/base-model.ts]: This interface is wierd. Please explain it face to face

File: src/app/view-project/view-project.component.html [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+
+
+  </tr>
+  </thead>
+  <tbody>
+  <ng-container *ngFor="let project of projects | async; index as i">
+    <tr>
+      <th scope="row">{{i + 1}}</th>
+      <td>
+        <svg (click)="deleteProject(project.name)" xmlns="http://www.w3.org/2000/svg" width="16" height="16"
```
Comment (yardenfi) [inline src/app/view-project/view-project.component.html]: Do you have a reason why do you need the svg embedded inside your component? if not, use font awesome or an image to display this.

File: src/app/view-project/view-project.component.html [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+             viewBox="0 0 16 16">
+          <path
+            d="M12.146.146a.5.5 0 0 1 .708 0l3 3a.5.5 0 0 1 0 .708l-10 10a.5.5 0 0 1-.168.11l-5 2a.5.5 0 0 1-.65-.65l2-5a.5.5 0 0 1 .11-.168l10-10zM11.207 2.5L13.5 4.793 14.793 3.5 12.5 1.207 11.207 2.5zm1.586 3L10.5 3.207 4 9.707V10h.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.5h.293l6.5-6.5zm-9.761 5.175l-.106.106-1.528 3.821 3.821-1.528.106-.106A.5.5 0 0 1 5 12.5V12h-.5a.5.5 0 0 1-.5-.5V11h-.5a.5.5 0 0 1-.468-.325z"/>
+        </svg>
+      </td>
+      <td>{{project.name}}</td>
+      <td>{{project.api}}</td>
+      <td>{{project.description}}</td>
+      <td>
+        <select autocomplete="on" class="form-resources" aria-label="Default select example" (change)="openDialogResource(i)"
```
Comment (yardenfi) [inline src/app/view-project/view-project.component.html]: It's wierd to show empty select when there are no resources created. Find a solution for this...

File: src/app/view-project/view-project.component.ts [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+        this.projects = this.projectService.getProjects();
+      }
+    });
+  }
+  openDialogResource(projectIndex: number): void {
+    if (!this.editResource){
+      return;
+    }
+    const resource: Resource = {
+      name: this.editResource[0],
```
Comment (yardenfi) [inline src/app/view-project/view-project.component.ts]: editResource is a string, I don't understand this...

File: src/app/view-project/view-project.component.ts [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+  deleteProject(name: string): void {
+    this.projectService.deleteProject(name);
+    this.projects = this.projectService.getProjects();
+  }
+  deleteResource(projectIndex: number, name: string, api: string, description: string, action: string): void {
+    const resource: Resource = {
+      name,
+      api,
+      description,
+     action
```
Comment (yardenfi) [inline src/app/view-project/view-project.component.ts]:
```suggestion
      action
```

File: src/app/view-project/view-project.component.ts [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+  templateUrl: './view-project.component.html',
+  styleUrls: ['./view-project.component.css']
+})
+export class ViewProjectComponent implements OnInit {
+  public projects: Observable<Project[]>;
+  public editResource: string;
+  constructor(private projectService: ProjectService, private resourceService: ResourceService, public dialog: DialogService) { }
+
+  ngOnInit(): void {
+    this.projects = this.projectService.getProjects();
```
Comment (yardenfi) [inline src/app/view-project/view-project.component.ts]: Observable variable names are usually marked with $ inside.
```suggestion
    this.projects$ = this.projectService.getProjects();
```

Comment (yardenfi) [inline src/app/view-project/view-project.component.ts]: You aren't calling unsubscribe anywhere.
Suggestion - read about rxjs Subscription.

File: src/services/project-service/project.service.ts [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+import {Observable} from 'rxjs';
+
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ProjectService {
+  constructor(private localStorageService: LocalStorageService,  private popUpService: PopupService) { }
+
+  private taskExists(task: string, currentTasks: Project[]): Project {
```
Comment (yardenfi) [inline src/services/project-service/project.service.ts]: What is a task?

File: src/services/project-service/project.service.ts [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+  public getProjects(): Observable<Project[]> {
+    return this.localStorageService.getStorage(PROJECT_KEY);
+  }
+  addProject(task: Project): void {
+    this.getProjects().subscribe(currentTasks => {
+      if (!this.taskExists(task.name, currentTasks)) {
+        currentTasks.push(task);
+        this.localStorageService.setStorage(PROJECT_KEY, currentTasks);
+        this.popUpService.showSuccess(`Successfully created ${task.name}`, 'Project');
+        return;
```
Comment (yardenfi) [inline src/services/project-service/project.service.ts]: No need for this. It is the default behaviour...

File: src/services/project-service/project.service.ts [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+export class ProjectService {
+  constructor(private localStorageService: LocalStorageService,  private popUpService: PopupService) { }
+
+  private taskExists(task: string, currentTasks: Project[]): Project {
+    for (const project of currentTasks) {
+      if (task === project.name) {
+        return project;
+      }
+    }
+    return null;
```
Comment (yardenfi) [inline src/services/project-service/project.service.ts]: Isn't this the default?

File: src/services/project-service/project.service.ts [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+        this.localStorageService.setStorage(PROJECT_KEY, currentTasks);
+        this.popUpService.showSuccess(`Successfully deleted ${name}`, 'Project');
+        return;
+      }
+      this.popUpService.showFailure(`Failed to delete ${name}`, 'Project');
+    });
+  }
+
+  editProject(name: string, editedProject: Project): void {
+    if (editedProject.name === '') {
```
Comment (yardenfi) [inline src/services/project-service/project.service.ts:54]: This seems like a validator. Read about angular forms:
https://angular.io/guide/forms-overview

File: src/app/create-project/create-project.component.html [hunk 25]
```diff
@@ -0,0 +1,40 @@
+<div class="forming">
+  <h2> Create a new project</h2>
+
+  <form class="row g-3">
+    <div class="col-md-6">
+      <label for="name" class="form-label">Project Name</label>
+      <input #projectName required minlength="3" type="text" class="form-control" ngModel name="name" #name="ngModel"
```
Comment (yardenfi) [inline src/app/create-project/create-project.component.html]: Refactor this form. Either use Reactive or Template driven forms, but do it properly

File: src/services/project-service/project.service.ts [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+      if (task === project.name) {
+        return project;
+      }
+    }
+    return null;
+  }
+  public getProjects(): Observable<Project[]> {
+    return this.localStorageService.getStorage(PROJECT_KEY);
+  }
+  addProject(task: Project): void {
```
Comment (yardenfi) [inline src/services/project-service/project.service.ts]: What happens if the server has returned an error? How can you show notification when the project has been created?
These are the reasons we always return an observable from the service. This way, the component can react to the result. Also, subscribing requires unsubscribing, usually the component does both...

File: src/services/resource-service/resource.service.ts [hunk 27]
```diff
@@ -0,0 +1,75 @@
...
+      const task = this.getResource(currentTasks[projectIndex], resource.name);
+      if (!task) {
+        currentTasks[projectIndex].resources.push(resource);
+        this.localStorageService.setStorage(PROJECT_KEY, currentTasks);
+        this.popUpService.showSuccess(`Successfully created Resource ${resource.name} in Project ${currentTasks[projectIndex].name}`,
+          'Resource');
+        return;
+      }
+      else {
+        this.popUpService.showFailure(`Failed to create Resource ${resource.name} in Project ${currentTasks[projectIndex].name}`,
```
Comment (yardenfi) [inline src/services/resource-service/resource.service.ts]: services are meant to encapsulate logic and server side handling, leave notifications and other user related actions to the components

File: src/services/storage-service/local-storage-service.service.ts [hunk 28]
```diff
@@ -0,0 +1,20 @@
+import {Inject, Injectable} from '@angular/core';
+import {LOCAL_STORAGE, StorageService} from 'ngx-webstorage-service';
+import {Observable, of} from 'rxjs';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class LocalStorageService {
+  constructor(@Inject(LOCAL_STORAGE) private storage: StorageService) { }
+  public getStorage(key: string): Observable<any> {
```
Comment (yardenfi) [inline src/services/storage-service/local-storage-service.service.ts:12]: face to face

**Comment by yardenfi** [comment]
Fix proportions of margins and buttons.
image

File: src/app/create-project/create-project.component.html [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+  <h2> Create a new project</h2>
+  <form>
+    <div class="form-group">
+      <label for="name" class="form-label">Project Name</label>
+      <input #projectName required minlength="3" type="text" class="form-control" ngModel name="name" #name="ngModel"
+             id="name" placeholder="My New Project">
+    </div>
+    <div class="form-group">
+      <label for="api" class="form-label">API base url</label>
+      <input #api type="url" class="form-control" id="api" placeholder="/isAlive">
```
Comment (yardenfi) [inline src/app/create-project/create-project.component.html]: remove the word "api"

Comment (yardenfi) [inline src/app/create-project/create-project.component.html]: This should be a full url. It needs make sure the value starts with http or https, then ://, then at least one letter...

File: src/app/create-project/create-project.component.html [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+    </div>
+    <div class="form-group">
+      <label for="general" class="form-label">General Description</label>
+      <input #desc type="text" class="form-control" id="general"
+             placeholder="Checks if the url sends a ping back">
+    </div>
+    <div class="form-group">
+      <button
+        *ngIf="!(name.invalid && (name.dirty || name.touched) && (name.errors.minlength || name.errors.required)) && projectName.value.length !== 0"
```
Comment (yardenfi) [inline src/app/create-project/create-project.component.html]: "ngIf"s shouldn't be so long, if you need conditions that are long, move this to a function

File: src/app/create-project/create-project.component.html [hunk 31]
```diff
@@ -0,0 +1,35 @@
...
+        (click)="addProject([projectName.value, api.value, desc.value])" type="submit" class="btn btn-primary"><a
+        class="btn btn-primary" href="/viewProjects">
```
Comment (yardenfi) [inline src/app/create-project/create-project.component.html]: We don't use href in angular, can you tell why? (hint: look for SPA)

File: src/app/create-project/create-project.component.html [hunk 32]
```diff
@@ -0,0 +1,35 @@
...
+      <label for="api" class="form-label">API base url</label>
+      <input #api type="url" class="form-control" id="api" placeholder="/isAlive">
+    </div>
+    <div class="form-group">
+      <label for="general" class="form-label">General Description</label>
+      <input #desc type="text" class="form-control" id="general"
+             placeholder="Checks if the url sends a ping back">
+    </div>
+    <div class="form-group">
+      <button
```
Comment (yardenfi) [inline src/app/create-project/create-project.component.html]: The submit button should always appear on screen, you can use [disabled] attribute to prevent the user from clicking the button while the form is invalid

File: src/app/create-project/create-project.component.html [hunk 33]
```diff
@@ -0,0 +1,35 @@
...
+        *ngIf="!(name.invalid && (name.dirty || name.touched) && (name.errors.minlength || name.errors.required)) && projectName.value.length !== 0"
+        (click)="addProject([projectName.value, api.value, desc.value])" type="submit" class="btn btn-primary"><a
+        class="btn btn-primary" href="/viewProjects">
+        Create </a></button>
+    </div>
+    <div *ngIf="name.invalid && (name.dirty || name.touched)"
+         class="alert alert-danger">
+      <div *ngIf="name.errors.minlength">
+        Project name must be at least 3 characters long.
```
Comment (yardenfi) [inline src/app/create-project/create-project.component.html]: a. form validation should appear below the respective control, not in the end of the form
b. invalid and touched controls should have indication that something is wrong, like a red border for example...
tip: Invalid controls have classes that indicate it. Use them

File: src/app/create-project/create-project.component.ts [hunk 34]
```diff
@@ -0,0 +1,35 @@
...
+@Component({
+  selector: 'app-project',
+  templateUrl: './create-project.component.html',
+  styleUrls: ['./create-project.component.css']
+})
+export class CreateProjectComponent implements OnInit {
+
+  constructor(private projectService: ProjectService, private toastr: ToastrService) { }
+
+  ngOnInit(): void {
```
Comment (yardenfi) [inline src/app/create-project/create-project.component.ts]: remove

File: src/app/create-project/create-project.component.html (hunk 29 above)
Comment (yardenfi) [inline src/app/create-project/create-project.component.html]: This placeholder is wrong, It should be a full url...

File: src/app/create-project/create-project.component.ts [hunk 35]
```diff
@@ -0,0 +1,35 @@
...
+      resources: []
+    };
+
+    const projectAdd = this.projectService.addProject(project).subscribe(
+      projects => {
+        this.toastr.success(`Successfully created project ${project.name}`, 'Project');
+      },
+          err => this.toastr.error(err, 'Project')
+    );
+    projectAdd.unsubscribe();
```
Comment (yardenfi) [inline src/app/create-project/create-project.component.ts]: It's great you've added unsubscribe, but there are two problems with it:
a. This won't work if the observable was actually async, as the unsubscribe would have been called before the response could be returned
b. this is unnecessary, as this observable is completed.

File: src/app/create-resource/create-resource.component.ts [hunk 36]
```diff
@@ -0,0 +1,44 @@
...
+  public projectIndex: number;
+  public projectName: string;
+  constructor(private activatedRoute: ActivatedRoute, private projectService: ProjectService,
+              private resourceService: ResourceService, private toastr: ToastrService) { }
+
+  ngOnInit(): void {
+    this.activatedRoute.queryParams.subscribe((params: Params) => {
+      this.projectIndex = params.projectID;
+      this.projectService.getProjects().subscribe(projects => {
+        this.projectName = projects[this.projectIndex].name;
```
Comment (yardenfi) [inline src/app/create-resource/create-resource.component.ts]: You shouldn't rely on the index of this array. Also, We tend not to rely on DB ids, but on uuid or some other unique form of identifier, such as names.

File: src/app/create-resource/create-resource.component.html [hunk 37]
```diff
@@ -0,0 +1,38 @@
...
+    </div>
+    <div class="form-group">
+      <label for="Path" class="form-label"> Path </label>
+      <input #path type="text" class="form-control" id="Path" placeholder="/delete">
+    </div>
+    <div class="form-group">
+      <button *ngIf= "!(name.invalid && (name.dirty || name.touched) && (name.errors.minlength || name.errors.required)) && resourceName.value.length !== 0"
+              (click)="addResource([resourceName.value, path.value, desc.value, rest.value])"  type="submit" class="btn btn-primary">
+        <a
+          class="btn btn-primary" href="/viewResources?projectID={{this.projectIndex}}">
```
Comment (yardenfi) [inline src/app/create-resource/create-resource.component.html]: again, avoid using href

File: src/app/create-resource/create-resource.component.html [hunk 38]
```diff
@@ -0,0 +1,38 @@
...
+      <label for="projectName" class="form-label"> Project Name </label>
+      <input type="text" disabled id="projectName" class="form-control" placeholder="{{projectName}}">
+    </div>
+    <div class="form-group">
+      <label for="name" class="form-label">Resource Name</label>
+      <input #resourceName placeholder="My New Resource" required minlength="3" type="text" class="form-control" ngModel name="name" #name="ngModel" id="name">
+    </div>
+    <div class="form-group">
+      <label for="rest" class="form-label"> Action </label>
+      <input #rest type="url" class="form-control" id="rest" placeholder="DELETE">
```
Comment (yardenfi) [inline src/app/create-resource/create-resource.component.html]: The user shouldn't be able to insert here whatever he\she likes, there is a closed list of options that could be set as value here. Which type of input would be more appropriate?

File: src/app/create-resource/create-resource.component.html [hunk 39]
```diff
@@ -0,0 +1,38 @@
...
+    </div>
+    <div class="form-group">
+      <label for="description" class="form-label">General Description</label>
+      <input #desc type="text" class="form-control" id="description" placeholder="deletes an object">
```
Comment (yardenfi) [inline src/app/create-resource/create-resource.component.html]: deletes an object?

File: src/app/edit-project/edit-project.component.ts [hunk 40]
```diff
@@ -0,0 +1,44 @@
...
+    this.onClose.next(project);
+    this.ref.hide();
+  }
+
+  public onCancel(): void {
+    this.onClose.next(null);
+    this.ref.hide();
+  }
+
+
```
Comment (yardenfi) [inline src/app/edit-project/edit-project.component.ts]: remove blank lines

File: src/app/edit-resource/edit-resource.component.html [hunk 41]
```diff
@@ -0,0 +1,31 @@
...
+      </div>
+      <div class="modal-body">
+        <form class="row g-3">
+          <div class="col-12">
+            <label class="form-label">Entity Name: </label>
+            <input name="name" type="text" [(ngModel)]="name">
+          </div>
+          <div class="col-12">
+            <label class="form-label">API url: </label>
+            <input name="api" type="text" [(ngModel)]="apiSuffix">
```
Comment (yardenfi) [inline src/app/edit-resource/edit-resource.component.html]: it's not api suffix, its a route

File: src/app/edit-resource/edit-resource.component.html [hunk 42]
```diff
@@ -0,0 +1,31 @@
+  <div class="modal-dialog" role="document">
+    <div class="modal-content">
+      <div class="modal-header">
+        <h5>Edit Resource {{name}}</h5>
+      </div>
+      <div class="modal-body">
+        <form class="row g-3">
+          <div class="col-12">
+            <label class="form-label">Entity Name: </label>
```
Comment (yardenfi) [inline src/app/edit-resource/edit-resource.component.html]: Resource name, It's not an entity

File: src/app/edit-resource/edit-resource.component.html [hunk 43]
```diff
@@ -0,0 +1,31 @@
+  <div class="modal-dialog" role="document">
```
Comment (yardenfi) [inline src/app/edit-resource/edit-resource.component.html]: What is the difference between edit and create resource components? Can they be merged or generalized?

File: src/app/edit-resource/edit-resource.component.ts [hunk 44]
```diff
@@ -0,0 +1,42 @@
...
+      apiSuffix: this.apiSuffix,
+      description: this.description,
+      method: this.method,
+    };
+
+    this.onClose.next(resource);
+    this.ref.hide();
+  }
+
+  public onCancel(): void {
```
Comment (yardenfi) [inline src/app/edit-resource/edit-resource.component.ts]: seems like your modals have a common interface, maybe add an abstraction?

**Comment by yardenfi** [comment]
Your site isn't really responsive because you use fixed width and height to describe things.
Find a way to fix this with a combination of flex, bootstrap containers and media queries

File: src/app/header/header.component.css [hunk 45]
```diff
@@ -0,0 +1,3 @@
+.header {
+  width: 120em;
```
Comment (yardenfi) [inline src/app/header/header.component.css]: this seems wrong...

File: src/app/header/header.component.html [hunk 46]
```diff
@@ -0,0 +1,15 @@
+<nav class="header navbar navbar-expand navbar-dark bg-primary">
+  <div class="container-fluid">
```
Comment (yardenfi) [inline src/app/header/header.component.html:2]: Why is this needed? you don't have any rows or columns

File: src/app/interfaces/resource.ts [hunk 47]
```diff
@@ -0,0 +1,6 @@
+export interface Resource {
+  name: string;
+  apiSuffix: string;
```
Comment (yardenfi) [inline src/app/interfaces/resource.ts]: route, not apiSuffix

File: src/app/view-project/view-project.component.css [hunk 48]
```diff
@@ -0,0 +1,27 @@
+h1 {
```
Comment (yardenfi) [inline src/app/view-project/view-project.component.css]: too broad selector

File: src/app/view-project/view-project.component.css [hunk 49]
```diff
@@ -0,0 +1,27 @@
...
+  margin-left: 20em;
+}
+
+.inner-spacing {
+  padding: 0.5em;
```
Comment (yardenfi) [inline src/app/view-project/view-project.component.css]: you can use p-1
read about bootstrap utility classes:
https://getbootstrap.com/docs/4.0/utilities/spacing/

File: src/app/view-project/view-project.component.css [hunk 50]
```diff
@@ -0,0 +1,27 @@
...
+
+.table-wrapper {
+  max-height: 20em;
+  overflow-y: auto;
+  width: 40em;
+  margin-left: 40em;
+  overflow-x: hidden;
+
+}
+
```
Comment (yardenfi) [inline src/app/view-project/view-project.component.css]: too many spaces

File: src/app/view-project/view-project.component.html [hunk 51]
```diff
@@ -0,0 +1,53 @@
+<h1> Your Projects </h1>
+<div class="table-wrapper">
+  <form class="form-inline my-2 my-lg-0">
+    <input #searchValue cha class="form-control" type="text" placeholder="Search" aria-label="Search">
+  </form>
+<table class="table">
+
```
Comment (yardenfi) [inline src/app/view-project/view-project.component.html]: remove spaces

File: src/app/view-project/view-project.component.html [hunk 52]
```diff
@@ -0,0 +1,53 @@
...
+  </tr>
+  </thead>
+  <tbody>
+  <ng-container *ngFor="let project of projects; index as i">
+    <tr *ngIf="project.name.startsWith(searchValue.value)">
+      <th scope="row">{{i + 1}}</th>
+      <td>
+        <svg (click)="deleteProject(project.name)" xmlns="http://www.w3.org/2000/svg" width="16" height="16"
+             fill="currentColor" class="bi bi-trash" viewBox="0 0 16 16">
+          <path
```
Comment (yardenfi) [inline src/app/view-project/view-project.component.html]: replace with proper image

File: src/app/view-resource/view-resource.component.ts [hunk 53]
```diff
@@ -0,0 +1,60 @@
...
+export class ViewResourceComponent implements OnInit {
+  public resources: Resource[];
+  public projectIndex: number;
+  public projectName; string;
+
+  constructor(private activatedRoute: ActivatedRoute, private projectService: ProjectService,
+              private resourceService: ResourceService, public modalService: BsModalService, private toastr: ToastrService) { }
+
+  ngOnInit(): void {
+    let innerSubscription = of().subscribe();
```
Comment (yardenfi) [inline src/app/view-resource/view-resource.component.ts]: a. you can just use new Subscription()
b. instead of this wierdness, merge the observables of queryparams and getProject to one observable. That way you can unsibscribe to them both, and also add error handling to both etc...

File: src/app/view-resource/view-resource.component.ts [hunk 54]
```diff
@@ -0,0 +1,60 @@
...
+        this.resources = projects[this.projectIndex].resources;
+        this.projectName = projects[this.projectIndex].name;
+      });
+    }).unsubscribe();
+    innerSubscription.unsubscribe();
+  }
+
+  openDialogResource(resource: Resource): void {
+    let innerSubscription = of().subscribe();
+    const dialogRef = this.modalService.show<EditResourceComponent>(EditResourceComponent, {initialState: resource});
```
Comment (yardenfi) [inline src/app/view-resource/view-resource.component.ts]: same about the observables here

File: src/app/view-resource/view-resource.component.ts [hunk 55]
```diff
@@ -0,0 +1,60 @@
...
+    innerSubscription.unsubscribe();
+  }
+
+  deleteResource(resource: Resource): void {
+    this.resourceService.deleteResource(this.projectIndex, resource).subscribe(projects => {
+      this.toastr.success(`Successfully deleted resource ${resource.name} in project ${projects[this.projectIndex].name}`, 'Resource');
+      this.resources = projects[this.projectIndex].resources;
+    },
+      err => this.toastr.error(err, 'Resource')
+    ).unsubscribe();
```
Comment (yardenfi) [inline src/app/view-resource/view-resource.component.ts]: Don't unsubscribe immediately after subscribing, I;v already wrote about it...

File: src/services/project-service/project.service.ts [hunk 56]
```diff
@@ -0,0 +1,73 @@
...
+
+@Injectable({
+  providedIn: 'root'
+})
+
+export class ProjectService {
+  constructor(private localStorageService: LocalStorageService) { }
+
+  private projectExists(proj: string, currentProjects: Project[]): Project {
+    for (const project of currentProjects) {
```
Comment (yardenfi) [inline src/services/project-service/project.service.ts]:
```suggestion
    return currentProjects.some(currentProject => currentProject.name == proj)
```

File: src/services/project-service/project.service.ts [hunk 57]
```diff
@@ -0,0 +1,73 @@
...
+import {tap} from 'rxjs/operators';
+
+@Injectable({
+  providedIn: 'root'
+})
+
+export class ProjectService {
+  constructor(private localStorageService: LocalStorageService) { }
+
+  private projectExists(proj: string, currentProjects: Project[]): Project {
```
Comment (yardenfi) [inline src/services/project-service/project.service.ts]: proj is wrong here, if I understand it correctly, its projectName

File: src/services/project-service/project.service.ts [hunk 58]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  addProject(proj: Project): Observable<Project[]> {
+    return this.getProjects().pipe(
+      tap(currentProjects => {
+      if (!this.projectExists(proj.name, currentProjects)) {
+        currentProjects.push(proj);
+        this.localStorageService.setStorage(PROJECT_KEY, currentProjects);
+      } else {
+        return throwError(`Failed to create ${proj.name}`);
```
Comment (yardenfi) [inline src/services/project-service/project.service.ts]: Does this work?

File: src/services/project-service/project.service.ts (hunk 24 above)
Comment (yardenfi) [reply src/services/project-service/project.service.ts:54]: Its still here

File: src/services/resource-service/resource.service.ts [hunk 59]
```diff
@@ -0,0 +1,76 @@
...
+import {Project, PROJECT_KEY} from '../../app/interfaces/project';
+import {Observable, throwError} from 'rxjs';
+import {Resource} from '../../app/interfaces/resource';
+import {ProjectService} from '../project-service/project.service';
+import {tap} from 'rxjs/operators';
+
+@Injectable({
+  providedIn: 'root'
+})
+
```
Comment (yardenfi) [inline src/services/resource-service/resource.service.ts]: remove blank lines in this file

File: src/services/resource-service/resource.service.ts [hunk 60]
```diff
@@ -0,0 +1,76 @@
...
+export class ResourceService {
+
+  constructor(private localStorageService: LocalStorageService, private projectService: ProjectService) { }
+
+  private getResource(project: Project, name: string): Resource {
+    for (const source of project.resources) {
```
Comment (yardenfi) [inline src/services/resource-service/resource.service.ts]: use Array.find

File: src/services/resource-service/resource.service.ts [hunk 61]
```diff
@@ -0,0 +1,76 @@
...
+        return throwError(`Failed to create Resource ${resource.name} in Project ${currentResources[projectIndex].name}`);
+        }
+      }
+    ));
+  }
+
+  deleteResource(projectIndex: number, resource: Resource): Observable<Project[]> {
+    return this.projectService.getProjects().pipe(
+      tap(currentResources => {
+      const foundResource = this.getResource(currentResources[projectIndex], resource.name);
```
Comment (yardenfi) [inline src/services/resource-service/resource.service.ts]: fix indentation

**Comment by yardenfi** [comment]
image
fix the css of your modals

File: src/services/project-service/project.service.ts (hunk 58 above)
Comment (barvered96) [reply src/services/project-service/project.service.ts]: yes
